generators/counter.py

```diff
@@ -19,36 +22,32 @@ def __init__(self):
         super().__init__()
         self.cur_count = 0
         self.count_lock = Lock()
+        self._lasti = 0
+        self.range = None
 
     def configure(self, context):
         super().configure(context)
-        # Initialize our count to the starting value
-        self.cur_count = self.attr_value.start
+        self._lasti = 0
+        self.range = range(self.attr_value.start, self.attr_value.end + 1,
+                           self.attr_value.step)
 
     def generate_signals(self, n=1):
-        return [self._get_next_signal() for i in range(n)]
-
-    def _get_next_signal(self):
-        """ Get the next signal we want to emit.
-
-        This will update the cur_count and return a Signal with the current
-        count at the time the function was called.
-        """
+        # bring variables into local space for minor speedup
         with self.count_lock:
-            # store the current count, then increment it
-            my_count = self.cur_count
-            self._bump_count()
-
-        return Signal({
-            self.attr_name: my_count
-        })
-
-    def _bump_count(self):
-        """ Increment the current count, roll over if necessary.
-
-        Returns:
-            None: self.cur_count gets updated instead
-        """
-        self.cur_count += self.attr_value.step
-        if self.cur_count > self.attr_value.end:
-            self.cur_count = self.attr_value.start
+            S = Signal
+            myrange = self.range
+            rlen = len(myrange)
+
+            ranges = []
+            sigs = 0
+            # this is how much is left of the previously used range
+            lasti = self._lasti
+            while sigs < n:
```

Comment:
After looking at it more, I think the one thing that will be different is where the next iteration will pick up the counter. 

So with the current setup, outputs will look like this:
```
[1, 2, 3, 1, 2]
[3, 1, 2, 3, 1]
[2, 3, 1, 2, 3]
...
```

but on the cleaned up it would be like
```
[1, 2, 3, 1, 2]
[1, 2, 3, 1, 2]
[1, 2, 3, 1, 2]
...
```

Let's talk tomorrow about the implications of this.